Methods that return unsized types can be (uselessly) codegened

``` Rust
trait Foo {
    fn foo(&self) -> [u8];
}

fn foo(f: Option<&Foo>) {
    if let Some(f) = f {
        let _ = f.foo();
    }
}

fn main() { foo(None) }
```

This does generate code that calls `foo` in the vtable, but is not particularly useful. 

cc @nikomatsakis

This code is now rejected with:

```
error[E0161]: cannot move a value of type [u8]: the size of [u8] cannot be statically determined
 --> src/main.rs:7:17
  |
7 |         let _ = f.foo();
  |                 ^^^^^^^

error: aborting due to previous error
```

--- a/src/test/ui/unsized/return-unsized-from-trait-method.rs
+++ b/src/test/ui/unsized/return-unsized-from-trait-method.rs
@@ -1,0 +1,16 @@
+// ignore-tidy-linelength
+
+// regression test for #26376
+
+trait Foo {
+    fn foo(&self) -> [u8];
+}
+
+fn foo(f: Option<&dyn Foo>) {
+    if let Some(f) = f {
+        let _ = f.foo();
+        //~^ ERROR cannot move a value of type [u8]: the size of [u8] cannot be statically determined
+    }
+}
+
+fn main() { foo(None) }

--- a/src/test/ui/unsized/return-unsized-from-trait-method.stderr
+++ b/src/test/ui/unsized/return-unsized-from-trait-method.stderr
@@ -1,0 +1,9 @@
+error[E0161]: cannot move a value of type [u8]: the size of [u8] cannot be statically determined
+  --> $DIR/return-unsized-from-trait-method.rs:11:17
+   |
+LL |         let _ = f.foo();
+   |                 ^^^^^^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0161`.

